plenum/server/consensus/consensus_data_provider.py

```diff
@@ -75,70 +75,72 @@ def stable_checkpoint(self) -> int:
             raise LogicError('Unprepared pp must be stored in preprepared')
         self._preprepared.remove(pp)
         self._prepared.append(pp)
 
-    def free_batch(self, pp):
+    def clear_batch(self, pp: PrePrepare):
         """
-        When 3pc batch had committed, rejected or catchuped till, it removed from _prepared list
+        When 3pc batch processed, it removed from _prepared list
         """
-        if pp is None:
-            raise LogicError('pp does not exist')
-        if pp in self._preprepared:
-            raise LogicError('Prepared pp cannot be stored in preprepared')
-        if pp not in self._prepared:
-            raise LogicError('Prepared pp must be stored in prepared')
+        # TODO: move logic checks upper the functions stack
+        if pp in self._preprepared == pp not in self._prepared:
+            raise LogicError('Batch can be only in one condition while clearing ')
         self._prepared.remove(pp)
 
-    def free_all(self):
+    def clear_all_batches(self):
         """
-        When catchup finished, backups clear up their pre-prepares
+        Clear up all preprepared and prepared
         """
         self._prepared.clear()
         self._preprepared.clear()
 
-    def get_3pc(self):
-        return self.view_no, self.pp_seq_no
-
-    def set_3pc(self, key):
-        if len(key) != 2:
-            raise LogicError('3pc key must be a pair')
-        self.view_no = key[0]
-        self.pp_seq_no = key[1]
-
     def append_checkpoint(self, checkpoint: Checkpoint):
-        self._checkpoints[(checkpoint.seqNoStart, checkpoint.seqNoEnd)] = checkpoint
+        self._checkpoints.add(checkpoint)
 
-    def remove_checkpoint(self, key):
-        if key not in self._checkpoints:
-            raise LogicError('key for removal must be in checkpoints')
-        del self._checkpoints[key]
+    def remove_checkpoint(self, end_seq_no):
+        new_checkpoints = [c for c in self._checkpoints if c.seqNoEnd != end_seq_no]
+        if len(new_checkpoints) != len(self._checkpoints) - 1:
+            raise LogicError('One checkpoint needed to be removed')
+        self._checkpoints = new_checkpoints
 
     def remove_all_checkpoints(self):
+        # That function most probably redundant in PBFT approach,
+        # because according to paper, checkpoints cleared only when next stabilized.
+        # Avoid using it while implement other services.
         self._checkpoints.clear()
 
-    def set_stable_checkpoint(self, key):
-        if key not in self._checkpoints:
-            raise LogicError('key for stable checkpoint must be in checkpoints')
-        self._stable_checkpoint = key[1]
+    def set_stable_checkpoint(self, end_seq_no):
+        for c in self._checkpoints:
+            if c.seqNoEnd == end_seq_no:
+                self._stable_checkpoint = end_seq_no
+                break
+        else:
+            raise LogicError('Stable checkpoint must be in checkpoints')
```

Comment:
I would refrain from using `for .. else` statement - it's quite error prone. Also when we set stable checkpoint we should delete all checkpoints that went below it.

Reply:
fixed